# This file is part of HiGFXback

# requires
REQUIRES="cairo-build gdk-pixbuf-build gnome-backgrounds-build make-build"

PKG_CONFIG_PATH=/fb/share/pkgconfig pkg-config --exists --print-errors $REQUIRES || exit 1

# configure
rm -f kapow/kapow.o
rm -f png/star_and_ring
PKG_CONFIG_PATH=/fb/lib/pkgconfig sh ./Configure --with-backgroundsdir=/share/backgrounds --with-demosdatadir=/fb/share/cairo-demos --prefix=/fb

# build
make
make -C kapow
make -C png
make -C PS

# install
install -d $DESTDIR/fb/bin
install chart-demo $DESTDIR/fb/bin
install fish-demo $DESTDIR/fb/bin
install flowers-demo $DESTDIR/fb/bin
install gears-demo $DESTDIR/fb/bin
install gradient-demo $DESTDIR/fb/bin
install slideshow-demo $DESTDIR/fb/bin
install spinner-demo $DESTDIR/fb/bin
install spiral-demo $DESTDIR/fb/bin
install tiger-demo $DESTDIR/fb/bin
install kapow/kapow $DESTDIR/fb/bin
install png/bevels $DESTDIR/fb/bin
install png/caps_joins $DESTDIR/fb/bin
install png/hering $DESTDIR/fb/bin
install png/outline $DESTDIR/fb/bin
install png/snapping $DESTDIR/fb/bin
install png/spiral $DESTDIR/fb/bin
install png/spline-pipelines $DESTDIR/fb/bin
install png/splines_tolerance $DESTDIR/fb/bin
install png/star_and_ring $DESTDIR/fb/bin
install png/stars $DESTDIR/fb/bin
install png/text $DESTDIR/fb/bin
install png/text-rotate $DESTDIR/fb/bin
install PS/basket $DESTDIR/fb/bin
install -d $DESTDIR/fb/share/cairo-demos
install -m 644 fishbg.png $DESTDIR/fb/share/cairo-demos
install -m 644 fishstrip.png $DESTDIR/fb/share/cairo-demos

# build.pc
install -d $DESTDIR/fb/share/pkgconfig
cat > $DESTDIR/fb/share/pkgconfig/cairo-demos-build.pc << EOF
Name: cairo-demos
Version: 20120130
Description: A collection of Cairo demos and test programs
Requires: $REQUIRES

exec=\\
/fb/bin/basket \\
/fb/bin/bevels \\
/fb/bin/caps_joins \\
/fb/bin/chart-demo \\
/fb/bin/fish-demo \\
/fb/bin/flowers-demo \\
/fb/bin/gears-demo \\
/fb/bin/gradient-demo \\
/fb/bin/hering \\
/fb/bin/kapow \\
/fb/bin/outline \\
/fb/bin/slideshow-demo \\
/fb/bin/snapping \\
/fb/bin/spinner-demo \\
/fb/bin/spiral \\
/fb/bin/spiral-demo \\
/fb/bin/spline-pipelines \\
/fb/bin/splines_tolerance \\
/fb/bin/star_and_ring \\
/fb/bin/stars \\
/fb/bin/text \\
/fb/bin/text-rotate \\
/fb/bin/tiger-demo \\
/fb/share/cairo-demos/fishbg.png \\
/fb/share/cairo-demos/fishstrip.png
EOF

exit

# patch
--- cairo-demos-20120130.orig/Configure
+++ cairo-demos-20120130/Configure
@@ -0,0 +1,39 @@
+#!/bin/sh
+
+PREFIX=/usr/local
+
+DEMOSDATADIR=.
+
+BACKGROUNDSDIR=.
+
+while [ $# != 0 ]; do
+  case $1 in
+    --prefix=*) PREFIX=${1#*=};;
+    --with-demosdatadir=*) DEMOSDATADIR=${1#*=};;
+    --with-backgroundsdir=*) BACKGROUNDSDIR=${1#*=};;
+    *) echo "Unknown option: '$1'"; exit 1;;
+  esac
+  shift
+done
+
+sed -i "/^PREFIX/s|=.*|=$PREFIX|" Makefile
+sed -i "/^PREFIX/s|=.*|=$PREFIX|" kapow/Makefile
+sed -i "/^PREFIX/s|=.*|=$PREFIX|" png/Makefile
+sed -i "/^PREFIX/s|=.*|=$PREFIX|" PS/Makefile
+
+sed -i "/^DEMOSDATADIR/s|=.*|=$DEMOSDATADIR|" Makefile
+
+sed -i "/^BACKGROUNDSDIR/s|=.*|=$BACKGROUNDSDIR|" Makefile
+
+test $PKG_CONFIG || PKG_CONFIG=pkg-config
+if test $PKG_CONFIG_PATH; then
+  sed -i "/^PKG_CONFIG/s|=.*|=PKG_CONFIG_PATH=$PKG_CONFIG_PATH $PKG_CONFIG|" Makefile
+  sed -i "/^PKG_CONFIG/s|=.*|=PKG_CONFIG_PATH=$PKG_CONFIG_PATH $PKG_CONFIG|" kapow/Makefile
+  sed -i "/^PKG_CONFIG/s|=.*|=PKG_CONFIG_PATH=$PKG_CONFIG_PATH $PKG_CONFIG|" png/Makefile
+  sed -i "/^PKG_CONFIG/s|=.*|=PKG_CONFIG_PATH=$PKG_CONFIG_PATH $PKG_CONFIG|" PS/Makefile
+else
+  sed -i "/^PKG_CONFIG/s|=.*|=$PKG_CONFIG|" Makefile
+  sed -i "/^PKG_CONFIG/s|=.*|=$PKG_CONFIG|" kapow/Makefile
+  sed -i "/^PKG_CONFIG/s|=.*|=$PKG_CONFIG|" png/Makefile
+  sed -i "/^PKG_CONFIG/s|=.*|=$PKG_CONFIG|" PS/Makefile
+fi
--- cairo-demos-20120130.orig/Makefile
+++ cairo-demos-20120130/Makefile
@@ -1,6 +1,21 @@
-SOURCES:=demo.c xlib.c ximage.c
-REQUIRES:=cairo-xlib xext gdk-pixbuf-2.0
-DEFINES:=-DHAVE_XLIB=1 -DHAVE_XIMAGE=1
+PREFIX=/usr/local
+
+PKG_CONFIG=pkg-config
+
+DEMOSDATADIR=.
+BACKGROUNDSDIR=.
+
+DEFINES:=-DDEMOSDATADIR=\"$(DEMOSDATADIR)\" -DBACKGROUNDSDIR=\"$(BACKGROUNDSDIR)\"
+SOURCES:=demo.c
+REQUIRES:=cairo gdk-pixbuf-2.0
+
+FBDEV:=1
+ifneq ($(FBDEV),0)
+DEFINES+=-DHAVE_FBDEV=1
+SOURCES+=fbdev.c
+else
+DEFINES+=-DHAVE_FBDEV=0
+endif
 
 DRM:=0
 ifneq ($(DRM),0)
@@ -11,7 +26,16 @@
 DEFINES+=-DHAVE_DRM=0
 endif
 
-XCB:=$(shell pkg-config --exists cairo-xcb && echo 1 || echo 0)
+XLIB:=$(shell $(PKG_CONFIG) --exists cairo-xlib && echo 1 || echo 0)
+ifneq ($(XLIB),0)
+DEFINES+=-DHAVE_XLIB=1 -DHAVE_XIMAGE=1
+SOURCES+=xlib.c ximage.c
+REQUIRES+=cairo-xlib xext
+else
+DEFINES+=-DHAVE_XLIB=0 -DHAVE_XIMAGE=0
+endif
+
+XCB:=$(shell $(PKG_CONFIG) --exists cairo-xcb && echo 1 || echo 0)
 ifneq ($(XCB),0)
 DEFINES+=-DHAVE_XCB=1
 SOURCES+=xcb.c
@@ -20,7 +44,7 @@
 DEFINES+=-DHAVE_XCB=0
 endif
 
-GLX:=$(shell pkg-config --exists cairo-gl && echo 1 || echo 0)
+GLX:=$(shell $(PKG_CONFIG) --exists cairo-gl && echo 1 || echo 0)
 ifneq ($(GLX),0)
 DEFINES+=-DHAVE_GLX=1
 SOURCES+=glx.c
@@ -29,7 +53,7 @@
 DEFINES+=-DHAVE_GLX=0
 endif
 
-COGL:=$(shell pkg-config --exists cairo-cogl && echo 1 || echo 0)
+COGL:=$(shell $(PKG_CONFIG) --exists cairo-cogl && echo 1 || echo 0)
 ifneq ($(COGL),0)
 DEFINES+=-DHAVE_COGL=1
 SOURCES+=cogl.c
@@ -40,13 +64,13 @@
 
 all: spinner-demo spiral-demo slideshow-demo tiger-demo fish-demo flowers-demo gears-demo gradient-demo chart-demo
 
-ifeq ($(shell pkg-config --exists poppler-glib && echo 1), 1)
+ifeq ($(shell $(PKG_CONFIG) --exists poppler-glib && echo 1), 1)
 all: poppler-demo
 REQUIRES+=poppler-glib
 endif
 
-CFLAGS:=$(shell pkg-config --cflags $(REQUIRES)) -Wall -g3
-LIBS:=$(shell pkg-config --libs $(REQUIRES))
+CFLAGS:=$(shell $(PKG_CONFIG) --cflags $(REQUIRES)) -Wall -g3
+LIBS:=$(shell $(PKG_CONFIG) --libs $(REQUIRES)) -lm -Wl,-rpath,$(PREFIX)/lib
 
 spinner-demo: spinner-demo.c $(SOURCES) demo.h Makefile
 	$(CC) $(DEFINES) $(CFLAGS) -o $@ spinner-demo.c $(SOURCES) $(LIBS)
--- cairo-demos-20120130.orig/PS/Makefile
+++ cairo-demos-20120130/PS/Makefile
@@ -1,3 +1,7 @@
+PREFIX=/usr/local
+
+PKG_CONFIG=pkg-config
+
 EXAMPLES=basket
 
 # I'd like to put a bunch of compiler-specific warning flags here, but
@@ -9,16 +13,13 @@
 #
 # CFLAGS="-Wall -Wpointer-arith -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -fno-strict-aliasing" make
 
-MYCFLAGS=`pkg-config --cflags cairo libpng12`
-MYLDFLAGS=`pkg-config --libs cairo libpng12`
+MYCFLAGS=`$(PKG_CONFIG) --cflags cairo libpng`
+MYLDFLAGS=`$(PKG_CONFIG) --libs cairo libpng` -Wl,-rpath,$(PREFIX)/lib
 
 all: ${EXAMPLES}
 
-%.o: %.c
-	$(CC) -c $(CFLAGS) $(CPPFLAGS) ${MYCFLAGS} $< -o $@
-
 %: %.c
 	$(CC) $(CFLAGS) $(CPPFLAGS) ${MYCFLAGS} ${MYLDFLAGS} $^ -o $@
 
 clean:
-	rm -f ${EXAMPLES} *.o
+	rm -f ${EXAMPLES} *.o *.ps
--- cairo-demos-20120130.orig/chart-demo.c
+++ cairo-demos-20120130/chart-demo.c
@@ -243,7 +243,7 @@
 		for (n = 0; n < 5; n++)
 			chart_update(&c[n]);
 
-		if (benchmark < 0) {
+		if (benchmark < 0 && !getenv("CAIRO_DEMOS_QUIET")) {
 			delta = now.tv_sec - last_tty.tv_sec;
 			delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
 			frames++;
--- cairo-demos-20120130.orig/demo.c
+++ cairo-demos-20120130/demo.c
@@ -52,12 +52,14 @@
 {
 	if (strcmp (str, "none") == 0)
 		return CAIRO_ANTIALIAS_NONE;
+#if CAIRO_VERSION > CAIRO_VERSION_ENCODE(1,11,3)
 	else if (strcmp (str, "fast") == 0)
 		return CAIRO_ANTIALIAS_FAST;
 	else if (strcmp (str, "good") == 0)
 		return CAIRO_ANTIALIAS_GOOD;
 	else if (strcmp (str, "best") == 0)
 		return CAIRO_ANTIALIAS_BEST;
+#endif
 	else
 		return CAIRO_ANTIALIAS_DEFAULT;
 }
@@ -85,6 +87,7 @@
     struct device *device = 0;
     enum {
 	AUTO,
+	FBDEV,
 	XLIB,
 	XIMAGE,
 	XCB,
@@ -95,7 +98,9 @@
     int n;
 
     for (n = 1; n < argc; n++) {
-	if (strcmp (argv[n], "--xlib") == 0) {
+	if (strcmp (argv[n], "--fbdev") == 0) {
+	    backend = FBDEV;
+	} else if (strcmp (argv[n], "--xlib") == 0) {
 	    backend = XLIB;
 	} else if (strcmp (argv[n], "--xcb") == 0) {
 	    backend = XCB;
@@ -111,6 +116,8 @@
     }
 
     if (backend == AUTO) {
+	if (device == 0 && HAVE_FBDEV)
+	    device = fbdev_open (argc, argv);
 	if (device == 0 && HAVE_DRM)
 	    device = drm_open (argc, argv);
 	if (device == 0 && HAVE_XCB)
@@ -125,6 +132,9 @@
 	    device = ximage_open (argc, argv);
     } else switch (backend) {
     case AUTO:
+    case FBDEV:
+	device = fbdev_open (argc, argv);
+	break;
     case XLIB:
 	device = xlib_open (argc, argv);
 	break;
@@ -141,9 +151,7 @@
 	device = cogl_open (argc, argv);
 	break;
     case DRM:
-#if HAVE_DRM
 	device = drm_open (argc, argv);
-#endif
 	break;
     }
 
@@ -152,7 +160,10 @@
 	exit(1);
     }
 
-    printf("Using backend \"%s\"\n", device->name);
+    if (!getenv("CAIRO_DEMOS_QUIET")) {
+	printf("Using backend \"%s\"\n", device->name);
+    }
+
     return device;
 }
 
@@ -200,7 +211,7 @@
 				q[3] = p[2];
 #endif
 				p += 3;
-				q += 4;
+				q += 3;
 			}
 		} else {
 			guchar *end = p + 4 * width;
@@ -231,6 +242,8 @@
 		cairo_pixels += cairo_stride;
 	}
 
+	cairo_surface_mark_dirty(surface);
+
 	return surface;
 }
 
--- cairo-demos-20120130.orig/demo.h
+++ cairo-demos-20120130/demo.h
@@ -78,6 +78,12 @@
 static inline struct device *drm_open (int argc, char **argv) { return 0; }
 #endif
 
+#if HAVE_FBDEV
+struct device *fbdev_open (int argc, char **argv);
+#else
+static inline struct device *fbdev_open (int argc, char **argv) { return 0; }
+#endif
+
 cairo_surface_t *
 _cairo_image_surface_create_from_pixbuf(const GdkPixbuf *pixbuf);
 
--- cairo-demos-20120130.orig/fbdev.c
+++ cairo-demos-20120130/fbdev.c
@@ -0,0 +1,110 @@
+#include <cairo.h>
+
+#include "demo.h"
+
+#include <fcntl.h>
+#include <stdlib.h>
+#include <string.h>
+#include <linux/fb.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+
+struct fbdev_device {
+    struct device base;
+    struct framebuffer fb;
+
+    int double_buffered;
+};
+
+static void
+destroy (struct framebuffer *fb)
+{
+}
+
+static void
+show (struct framebuffer *fb)
+{
+    struct fbdev_device *device = (struct fbdev_device *) fb->device;
+    cairo_t *cr;
+
+    if (device->double_buffered) {
+        cr = cairo_create (device->base.scanout);
+        cairo_set_source_surface (cr, fb->surface, 0, 0);
+        cairo_paint (cr);
+        cairo_destroy (cr);
+    }
+}
+
+static struct framebuffer *
+get_fb (struct device *abstract_device)
+{
+    struct fbdev_device *device = (struct fbdev_device *) abstract_device;
+    return &device->fb;
+}
+
+struct device *
+fbdev_open (int argc, char **argv)
+{
+    struct fbdev_device *device;
+    int fd;
+    struct fb_var_screeninfo info;
+    unsigned char *data;
+    cairo_format_t format;
+    int i, x = 0, y = 0;
+
+    device = malloc(sizeof (struct fbdev_device));
+    device->base.name = "fbdev";
+    device->base.get_framebuffer = get_fb;
+
+    fd = open("/dev/fb0", O_RDWR);
+    ioctl (fd, FBIOGET_VSCREENINFO, &info);
+
+    device->base.width = info.xres;
+    device->base.height = info.yres;
+    device_get_size (argc, argv, &device->base.width, &device->base.height);
+
+    device->double_buffered = 1;
+    for (i = 1; i < argc; i++) {
+        if (strncmp (argv[i], "--position=", 11) == 0) {
+            sscanf (argv[i]+11, "%dx%d", &x, &y);
+        } else if (strcmp (argv[i], "--position") == 0) {
+            if (i + 1 < argc) sscanf (argv[++i], "%dx%d", &x, &y);
+        } else if (strcmp (argv[i], "--single") == 0) {
+            device->double_buffered = 0;
+        }
+    }
+
+    data = mmap (NULL,
+                 info.xres * info.yres * info.bits_per_pixel / 8,
+                 PROT_WRITE,
+                 MAP_SHARED,
+                 fd,
+                 0);
+
+    data = data + y * info.xres * info.bits_per_pixel / 8 + x * info.bits_per_pixel / 8;
+
+    if (info.bits_per_pixel == 24)
+        format = CAIRO_FORMAT_RGB24;
+    else
+        format = CAIRO_FORMAT_ARGB32;
+
+    device->base.scanout = cairo_image_surface_create_for_data (data,
+                                                                format,
+                                                                device->base.width,
+                                                                device->base.height,
+                                                                info.xres * info.bits_per_pixel / 8);
+
+    device->fb.device = &device->base;
+    if (device->double_buffered) {
+        device->fb.surface = cairo_surface_create_similar (device->base.scanout,
+                                                           CAIRO_CONTENT_COLOR,
+                                                           device->base.width,
+                                                           device->base.height);
+    } else {
+        device->fb.surface = cairo_surface_reference (device->base.scanout);
+    }
+    device->fb.show = show;
+    device->fb.destroy = destroy;
+
+    return &device->base;
+}
--- cairo-demos-20120130.orig/fish-demo.c
+++ cairo-demos-20120130/fish-demo.c
@@ -22,7 +22,7 @@
 	double sf;
 	cairo_t *cr;
 
-	image = cairo_image_surface_create_from_png("fishbg.png");
+	image = cairo_image_surface_create_from_png(DEMOSDATADIR"/fishbg.png");
 
 	height = cairo_image_surface_get_height(image);
 	width = cairo_image_surface_get_width(image);
@@ -57,7 +57,7 @@
 	int width, height, x, y;
 	cairo_t *cr;
 
-	image = cairo_image_surface_create_from_png("fishstrip.png");
+	image = cairo_image_surface_create_from_png(DEMOSDATADIR"/fishstrip.png");
 	width = cairo_image_surface_get_width(image);
 	height = cairo_image_surface_get_height(image);
 	surface = cairo_surface_create_similar(device->scanout,
@@ -241,7 +241,7 @@
 		fb->show (fb);
 		fb->destroy (fb);
 
-		if (benchmark < 0) {
+		if (benchmark < 0 && !getenv("CAIRO_DEMOS_QUIET")) {
 			delta = now.tv_sec - last_tty.tv_sec;
 			delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
 			frames++;
--- cairo-demos-20120130.orig/flowers-demo.c
+++ cairo-demos-20120130/flowers-demo.c
@@ -411,7 +411,7 @@
 
 		flowers_update(device->width, device->height);
 
-		if (benchmark < 0) {
+		if (benchmark < 0 && !getenv("CAIRO_DEMOS_QUIET")) {
 			delta = now.tv_sec - last_tty.tv_sec;
 			delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
 			frames++;
--- cairo-demos-20120130.orig/gears-demo.c
+++ cairo-demos-20120130/gears-demo.c
@@ -261,7 +261,7 @@
 		fb->show (fb);
 		fb->destroy (fb);
 
-		if (benchmark < 0) {
+		if (benchmark < 0 && !getenv("CAIRO_DEMOS_QUIET")) {
 			delta = now.tv_sec - last_tty.tv_sec;
 			delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
 			frames++;
--- cairo-demos-20120130.orig/gradient-demo.c
+++ cairo-demos-20120130/gradient-demo.c
@@ -200,7 +200,7 @@
 		fb->show (fb);
 		fb->destroy (fb);
 
-		if (benchmark < 0) {
+		if (benchmark < 0 && !getenv("CAIRO_DEMOS_QUIET")) {
 			delta = now.tv_sec - last_tty.tv_sec;
 			delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
 			frames++;
--- cairo-demos-20120130.orig/kapow/.gitignore
+++ cairo-demos-20120130/kapow/.gitignore
@@ -1 +1,2 @@
 kapow
+kapow.o
--- cairo-demos-20120130.orig/kapow/Makefile
+++ cairo-demos-20120130/kapow/Makefile
@@ -1,5 +1,9 @@
-MYCFLAGS= $(shell pkg-config --cflags cairo libpng) -Wall -g -Wpointer-arith -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -fno-strict-aliasing
-LDLIBS	= $(shell pkg-config --libs cairo libpng) -g
+PREFIX=/usr/local
+
+PKG_CONFIG=pkg-config
+
+MYCFLAGS= $(shell $(PKG_CONFIG) --cflags cairo libpng) -Wall -g -Wpointer-arith -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -fno-strict-aliasing
+LDLIBS	= $(shell $(PKG_CONFIG) --libs cairo libpng) -lm -Wl,-rpath,$(PREFIX)/lib
 
 all: kapow
 
@@ -9,4 +13,4 @@
 kapow : kapow.o
 
 clean :
-	rm kapow.o kapow
+	rm -f kapow.o kapow kapow.png
--- cairo-demos-20120130.orig/png/.gitignore
+++ cairo-demos-20120130/png/.gitignore
@@ -4,7 +4,9 @@
 outline
 snapping
 spiral
+spline-pipelines
 splines_tolerance
+star_and_ring
 stars
 text
 text-rotate
--- cairo-demos-20120130.orig/png/Makefile
+++ cairo-demos-20120130/png/Makefile
@@ -1,4 +1,8 @@
-EXAMPLES=bevels caps_joins hering outline spiral splines_tolerance snapping stars star_and_ring text text-rotate
+PREFIX=/usr/local
+
+PKG_CONFIG=pkg-config
+
+EXAMPLES=bevels caps_joins hering outline spiral spline-pipelines splines_tolerance snapping stars star_and_ring text text-rotate
 
 # I'd like to put a bunch of compiler-specific warning flags here, but
 # I don't know a good way to choose the right flags based on the
@@ -9,14 +13,11 @@
 #
 # CFLAGS="-Wall -Wpointer-arith -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -fno-strict-aliasing" make
 
-MYCFLAGS=`pkg-config --cflags cairo libpng12`
-MYLDFLAGS=`pkg-config --libs cairo libpng12`
+MYCFLAGS=`$(PKG_CONFIG) --cflags cairo libpng`
+MYLDFLAGS=`$(PKG_CONFIG) --libs cairo libpng` -Wl,-rpath,$(PREFIX)/lib
 
 all: $(EXAMPLES)
 
-%.o: %.c
-	$(CC) -c $(CFLAGS) $(CPPFLAGS) ${MYCFLAGS} $< -o $@
-
 %: %.c
 	$(CC) $(CFLAGS) $(CPPFLAGS) ${MYCFLAGS} ${MYLDFLAGS} $^ -o $@
 
--- cairo-demos-20120130.orig/png/snapping.c
+++ cairo-demos-20120130/png/snapping.c
@@ -43,6 +43,13 @@
 #include <cairo.h>
 #include <math.h>
 
+static inline int
+fast_floor (double x)
+{
+    int i = (int) x;
+    return i - (i > x);
+}
+
 #define WIDTH 175
 #define HEIGHT 175
 #define STRIDE (WIDTH * 4)
@@ -62,8 +69,8 @@
 {
     /* Convert to device space, round, then convert back to user space. */
     cairo_user_to_device (cr, x, y);
-    *x = floor (*x + 0.5);
-    *y = floor (*y + 0.5);
+    *x = fast_floor (*x + 0.5);
+    *y = fast_floor (*y + 0.5);
     cairo_device_to_user (cr, x, y);
 }
 
@@ -98,8 +105,8 @@
     y_offset = y_width_dev_2 - (int) (y_width_dev_2);
 
     cairo_user_to_device (cr, x, y);
-    *x = floor (*x + x_offset + 0.5);
-    *y = floor (*y + y_offset + 0.5);
+    *x = fast_floor (*x + x_offset + 0.5);
+    *y = fast_floor (*y + y_offset + 0.5);
     *x -= x_offset;
     *y -= y_offset;
     cairo_device_to_user (cr, x, y);
@@ -125,7 +132,7 @@
     /* If the line width is less than 1 then it will round to 0 and
      * disappear. Instead, we clamp it to 1.0, but we must preserve
      * its sign for the case of a reflecting transformation. */
-    x_width_snapped = floor (x_width + 0.5);
+    x_width_snapped = fast_floor (x_width + 0.5);
     if (fabs(x_width_snapped) < 1.0) {
 	if (x_width > 0)
 	    x_width_snapped = 1.0;
@@ -133,7 +140,7 @@
 	    x_width_snapped = -1.0;
     }
 
-    y_width_snapped = floor (y_width + 0.5);
+    y_width_snapped = fast_floor (y_width + 0.5);
     if (fabs(y_width_snapped) < 1.0) {
 	if (y_width > 0)
 	    y_width_snapped = 1.0;
--- cairo-demos-20120130.orig/slideshow-demo.c
+++ cairo-demos-20120130/slideshow-demo.c
@@ -142,7 +142,7 @@
 	}
 	closedir(dir);
 
-	if (count)
+	if (count && !getenv("CAIRO_DEMOS_QUIET"))
 		printf("Loaded %d images from %s\n", count, path);
 }
 
@@ -196,11 +196,13 @@
 		}
 	}
 	if (num_sources == 0)
-		load_sources("/usr/share/backgrounds", device,
+		load_sources(BACKGROUNDSDIR, device,
 			     device->get_framebuffer(device)->surface);
 
-	printf("Loaded %d images, %ld/%ld pixels in total\n",
-	       num_sources, (long)out_pixels, (long)in_pixels);
+	if (!getenv("CAIRO_DEMOS_QUIET"))
+		printf("Loaded %d images, %ld/%ld pixels in total\n",
+		       num_sources, (long)out_pixels, (long)in_pixels);
+
 	if (num_sources == 0)
 		return 0;
 
@@ -263,7 +265,7 @@
 		frames++;
 		delta = now.tv_sec - last_tty.tv_sec;
 		delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
-		if (delta >  5) {
+		if (delta >  5 && !getenv("CAIRO_DEMOS_QUIET")) {
 			printf("%.2f fps\n", frames/delta);
 			last_tty = now;
 			frames = 0;
--- cairo-demos-20120130.orig/spinner-demo.c
+++ cairo-demos-20120130/spinner-demo.c
@@ -149,7 +149,7 @@
 	}
 	closedir(dir);
 
-	if (count)
+	if (count && !getenv("CAIRO_DEMOS_QUIET"))
 		printf("Loaded %d images from %s\n", count, path);
 }
 
@@ -199,11 +199,13 @@
 		}
 	}
 	if (num_sources == 0)
-		load_sources("/usr/share/backgrounds", device,
+		load_sources(BACKGROUNDSDIR, device,
 			     device->get_framebuffer(device)->surface);
 
-	printf("Loaded %d images, %ld/%ld pixels in total\n",
-	       num_sources, (long)out_pixels, (long)in_pixels);
+	if (!getenv("CAIRO_DEMOS_QUIET"))
+		printf("Loaded %d images, %ld/%ld pixels in total\n",
+		       num_sources, (long)out_pixels, (long)in_pixels);
+
 	if (num_sources == 0)
 		return 0;
 
@@ -250,7 +252,7 @@
 		frames++;
 		delta = now.tv_sec - last_tty.tv_sec;
 		delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
-		if (delta >  5) {
+		if (delta >  5 && !getenv("CAIRO_DEMOS_QUIET")) {
 			printf("%.2f fps\n", frames/delta);
 			last_tty = now;
 			frames = 0;
--- cairo-demos-20120130.orig/spiral-demo.c
+++ cairo-demos-20120130/spiral-demo.c
@@ -124,7 +124,7 @@
 	}
 	closedir(dir);
 
-	if (count)
+	if (count && !getenv("CAIRO_DEMOS_QUIET"))
 		printf("Loaded %d images from %s\n", count, path);
 }
 
@@ -144,7 +144,7 @@
 int main(int argc, char **argv)
 {
 	struct device *device;
-	const char *path = "/usr/share/backgrounds";
+	const char *path = BACKGROUNDSDIR;
 	float sincos_lut[360];
 	struct timeval start, last_tty, last_fps, now;
 	int theta, frames, n;
@@ -160,6 +160,8 @@
 		if (strcmp (argv[n], "--images") == 0) {
 			path = argv[n+1];
 			n++;
+	    } else if (strcmp (argv[n], "--prescale") == 0) {
+		prescale = 1;
 	    } else if (strcmp (argv[n], "--hide-path") == 0) {
 		show_path = 0;
 	    } else if (strcmp (argv[n], "--hide-outline") == 0) {
@@ -181,8 +183,10 @@
 	}
 
 	load_sources(path, device->get_framebuffer(device)->surface);
-	printf("Loaded %d images, %ld/%ld pixels in total from %s\n",
-	       num_sources, (long)out_pixels, (long)in_pixels, path);
+
+	if (!getenv("CAIRO_DEMOS_QUIET"))
+		printf("Loaded %d images, %ld/%ld pixels in total from %s\n",
+		       num_sources, (long)out_pixels, (long)in_pixels, path);
 
 	gettimeofday(&start, 0); now = last_tty = last_fps = start;
 	frames = 0;
@@ -203,7 +207,6 @@
 
 		rotation = floor(50*delta);
 
-		cairo_set_operator(cr, CAIRO_OPERATOR_CLEAR);
 		cairo_paint(cr);
 
 		cairo_set_operator(cr, CAIRO_OPERATOR_OVER);
@@ -299,7 +302,7 @@
 		frames++;
 		delta = now.tv_sec - last_tty.tv_sec;
 		delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
-		if (delta >  5) {
+		if (delta >  5 && !getenv("CAIRO_DEMOS_QUIET")) {
 			printf("%.2f fps\n", frames/delta);
 			last_tty = now;
 			frames = 0;
--- cairo-demos-20120130.orig/tiger-demo.c
+++ cairo-demos-20120130/tiger-demo.c
@@ -15,6 +15,7 @@
  */
 
 #include <cairo.h>
+#include <stdlib.h>
 #include <sys/time.h>
 
 #include "demo.h"
@@ -78,7 +79,9 @@
 {
 	switch (antialias) {
 	case CAIRO_ANTIALIAS_NONE: return "none";
+#if CAIRO_VERSION > CAIRO_VERSION_ENCODE(1,11,3)
 	case CAIRO_ANTIALIAS_FAST: return "fast";
+#endif
 	case CAIRO_ANTIALIAS_GRAY: return "gray";
 	case CAIRO_ANTIALIAS_SUBPIXEL: return "subpixel";
 	default: return "best";
@@ -127,10 +130,10 @@
 		fb->destroy (fb);
 
 		++frame;
-		if (frame % 256 == 0)
+		if (frame % 192 == 0)
 			factor = 1./factor;
 
-		if (benchmark < 0) {
+		if (benchmark < 0 && !getenv("CAIRO_DEMOS_QUIET")) {
 			delta = now.tv_sec - last_tty.tv_sec;
 			delta += (now.tv_usec - last_tty.tv_usec)*1e-6;
 			frames++;

# source
https://github.com/caramelli/cairo-demos/archive/de0b06d10b7cd05bf6ef193a2bbc93d8eb94bbfc/cairo-demos-20120130.tar.gz
